File: embulk-core/src/main/java/org/embulk/spi/DecoderPlugin.java
package org.embulk.spi;

import org.embulk.config.TaskSource;
import org.embulk.config.ConfigSource;

public interface DecoderPlugin
{
    interface Control
    {
        void run(TaskSource taskSource);
    }

    void transaction(ConfigSource config, DecoderPlugin.Control control);

    FileInput open(TaskSource taskSource, FileInput input);
}

File: embulk-ruby/lib/embulk/decoder_plugin.rb
module Embulk

  require 'embulk/data_source'

  class DecoderPlugin
    def self.transaction(config, &control)
      raise NotImplementedError, "DecoderPlugin.transaction(config, &control) must be implemented"
    end

    # TODO

    # TODO new_java

    def self.from_java(java_class)
      JavaPlugin.ruby_adapter_class(java_class, DecoderPlugin, RubyAdapter)
    end

    module RubyAdapter
      module ClassMethods
      end
      # TODO
    end
  end

end
